Feature/asm translator

Floris497 opened the pull request:
Mostly the translator and a small part of the tokenizer/meta parser

- feature: new data types and structs for future work
- chore: can now detect for source comment line
- feat: work in progress to get parts working
- feature: now read lines form string intead of file
- chore: new header
- feat: read lines from string instead of file
- feat: start on meta parser
- feat: start on tokenizer
- feature: bytecode translator for all instructions
- feat: makefill for past 9 commits

merlijnve commented on srcs/asm/parser/input_parser.c:
```diff
@@ -99,6 +99,9 @@ static t_line_type	line_type(char *line)
 	if (line == NULL)
 		return (kUndefinedLine);
 
+	if (*line == '#')
```
line 168: switches zijn verboden

merlijnve reviewed (commented):
Meeste lijkt nog WIP, wel goed om even door te kunnen kijken.
Switches zijn wel norm, en let op dat je niet te lange functies vervelend moet gaan opbreken later.

Floris497 commented:
Nee klopt, switch is makkelijk met opzetten, die moet ik nog ff aanpassen

Floris497 commented:
Added a load of new commits, not only translator related.. please all review code and ask if anything is unclear

- chore: removed dead stores
- chore: proper use of strncmp
- feature: add ft_lstlen to libft
- feature: add error codes and place as t_index types
- chore: use the chars specified in op.h instead of magic
- feature: retrieve lines from memory loaded file
- chore: work in progress for meta parser
- feature: added tokenizer to project 
- chore: make headers equal
- feature: massively smaller translator functions
- feature: added missing tokens for indirect
- fix: comparing at full length
- fix: labels now correctly tokenized
- feature: added function for indirects
- feature: add new func put to memory in little endian
- feature: remove part count from translate functions
- fix: corrections for indirects and sizes for directs
- feature: add translator and translate to main
- chore: makefile now compiles project correctly
- fix: function correctly named
- feature: jumps now saved as index instead of pointer
- feature: correctly adding markers
- fix: prototype for put indirect and call
- feature: refactor error code struct
- feature: added linker and redone linking hints to save more info
- fix: the stupidest mistake ever!
- chore: more checks for malloc fail
- chore: refactor debugging code
- fix: better comparissons on labels
- fix: making code norm complient as much as possible

ItsWendell commented on srcs/asm/asm.h:
```diff
@@ -53,34 +73,72 @@ enum    e_line_type
 enum e_inst
```
Can we share this with the VM?

Floris497 replied:
After we merge can put move it to op.h

ItsWendell commented on srcs/asm/parser/input_parser.c:
```diff
@@ -12,43 +12,34 @@
 #include "input_parser.h"
+#include "tokenizer.h"
+
+static char *inst_strs[16] =
```
shoudn't we avoid global vars?

ItsWendell commented on srcs/lib/libft/ft_putmembe.c:
```diff
@@ -0,0 +1,29 @@
+	while (idx + 1 <= size)
+	{
+		mems[idx] =
+		((num & (0xFFL << (size - (idx + 1)) * 8)) >> (size - (idx + 1)) * 8);
```
WHaatt's happening?

Floris497 replied:
Oh don't ask.. this makes sure that whatever endian encoding is used by the CPU it will always write Big Endian to the file.. there is just no easy way to write this down comprehensively

ItsWendell replied:
Can we put that in a norm compliant comment above the function?

Floris497 replied:
Will do

ItsWendell reviewed (commented):
This is a lot but looks very clean, take a look at your // TODO comments and double check to see if you can remove them!
